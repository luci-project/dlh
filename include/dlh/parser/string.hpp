#pragma once

#include <dlh/container/vector.hpp>
#include <dlh/types.hpp>

namespace Parser {

bool string(unsigned long long & target, const char * value);

bool string(long long & target, const char * value);

bool string(unsigned long & target, const char * value);

bool string(long & target, const char * value);

bool string(unsigned int & target, const char * value);

bool string(int & target, const char * value);

bool string(unsigned short & target, const char * value);

bool string(short & target, const char * value);

bool string(unsigned char & target, const char * value);

bool string(char & target, const char * value);

bool string(bool & target, const char * value);

bool string(const char * & target, const char * value);

template<typename T>
bool string(Vector<T> & target, const char * value) {
	T tmp;
	bool r = string(tmp, value);
	target.push_back(tmp);
	return r;
}

}  // namespace Parse
