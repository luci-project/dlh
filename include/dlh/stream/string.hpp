// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#pragma once

#include <dlh/types.hpp>
#include <dlh/stream/buffer.hpp>

template<size_t BUFFERSZ>
class StringStream : public BufferStream {
 protected:
	char buffer[BUFFERSZ];

 public:
	/*! \brief Default constructor  */
	StringStream() : BufferStream(buffer, BUFFERSZ) {}

	/*! \brief Destructor */
	virtual ~StringStream() {}
};
