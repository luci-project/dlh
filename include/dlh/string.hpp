#pragma once

#include <dlh/types.hpp>
#include <dlh/container/vector.hpp>

namespace String {

/*! \brief Find the first occurrence of a character in a string
 * \param s string to search in
 * \param c character to find
 * \return Pointer to first occurrence of the character
 *         or to null byte at the end of the string if not found
 */
const char* find_or_end(const char *s, int c);

/*! \brief Find the first occurrence of a substring in a string
 * \param haystack string to search in
 * \param needle string to find
 * \return Pointer to first occurrence of the needle
 *         or to null byte at the end of the string if not found
 */
const char* find_or_end(const char *haystack, const char *needle);

/*! \brief Find the first occurrence of a character in a string
 * \param s string to search in
 * \param c character to find
 * \return Pointer to first occurrence of the character
 *         or to nullptr if not found
 */
const char* find(const char *s, int c);

/*! \brief Find the first occurrence of a substring in a string
 * \param haystack string to search in
 * \param needle string to find
 * \return Pointer to first occurrence of the needle
 *         or to nullptr if not found
 */
const char* find(const char *haystack, const char *needle);

/*! \brief Find the last occurrence of a character in a string
 * \param s string to search in
 * \param c character to find
 * \return Pointer to last occurrence of the character
 *         or to nullptr if not found
 */
const char* find_last(const char *s, int c);

/*! \brief Find the last occurrence of a substring in a string
 * \param haystack string to search in
 * \param needle string to find
 * \return Pointer to last occurrence of the needle
 *         or to nullptr if not found
 */
const char* find_last(const char *haystack, const char *needle);

/*! \brief Replace characters in a string
 * \param target pointer to string (target memory will be modified!)
 * \param from character to be replaced
 * \param to replacement character
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 */
char * replace_inplace(char *target, int from, int to, size_t max = SIZE_MAX);

/*! \brief Replace characters in a string
 * \param target pointer to string (string will be copied)
 * \param from character to be replaced
 * \param to replacement character
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 * \note Target pointer has to be freed after usage
 */
char * replace(const char *target, int from, int to, size_t max = SIZE_MAX);

/*! \brief Replace stubstrings in a string
 * \param target pointer to string  (target memory will be modified!)
 * \param from substring to be replaced
 * \param to replacement substring (must be equal or shorter than `from`)
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 */
char * replace_inplace(char *target, const char *from, const char * to, size_t max = SIZE_MAX);

/*! \brief Replace stubstrings in a string
 * \param target pointer to string (string will be copied)
 * \param from substring to be replaced
 * \param to replacement substring
 * \param max maximum number of replacements to be performed
 * \return Pointer to the target string
 * \note Target pointer has to be freed after usage
 */
char * replace(const char *target, const char *from, const char * to, size_t max = SIZE_MAX);

/*! \brief Compare two strings
 * \param s1 first string
 * \param s2 second string
 * \return an integer less than, equal to, or greater than zero if first string is found, respectively,
 *         to be less than, to match, or be greater than second string
 */
int compare(const char *s1, const char *s2);

/*! \brief Compare two strings
 * \param s1 first string
 * \param s2 second string
 * \param n number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first string are
 *          found, respectively, to be less than, to match, or be greater than second string
 */
int compare(const char *s1, const char *s2, size_t n);

/*! \brief Calculate the length of a string
 * \param s pointer to a string
 * \return number of bytes in the string
 */
size_t len(const char *s);

/*! \brief Copy the contents of a string
 * including the terminating null byte (`\0`)
 * \param dest destination string buffer
 * \param src source string buffer
 * \return a pointer to the destination string buffer
 * \note Beware of buffer overruns!
 */
char* copy(char * dest, const char * src);  //NOLINT

/*! \brief Copy the contents of a string up to a maximum length
 * or the terminating null byte (`\0`), whatever comes first.
 * \param dest destination string buffer
 * \param src source string buffer
 * \param n maximum number of bytes to copy
 * \return a pointer to the destination string buffer
 * \note If there is no null byte (`\0`) among the first `n` bytes, the destination will not be null-terminated!
 */
char* copy(char * dest, const char * src, size_t n);

/*! \brief Duplicate a string
 * \param s pointer to a string
 * \return pointer to a duplicated string allocated with malloc or `nullptr` if allocation failed.
 */
char* duplicate(const char *s);

/*! \brief Duplicate a string
 * \param s pointer to a string
 * \param n maximum length
 * \return pointer to a duplicated string (up to maximum length, always null-terminated),
 *         allocated with malloc or `nullptr` if allocation failed.
 */
char* duplicate(const char *s, size_t n);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string (target memory will be modified!)
 * \param delimiter character to split string (will be overwritten with NULL character)
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 */
Vector<const char *> split_inplace(char * source, int delimiter, size_t max = SIZE_MAX);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string
 * \param delimiter character to split string
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 * \note You have to free each element of the result vector!
 */
Vector<const char *> split(const char * source, int delimiter, size_t max = SIZE_MAX);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string (target memory will be modified!)
 * \param delimiter substring to split string
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 */
Vector<const char *> split_inplace(char * source, const char * delimiter, size_t max = SIZE_MAX);

/*! \brief Split a string by an delimiter
 * \param source pointer to a string
 * \param delimiter substring to split string
 * \param max maximum number of splits (hence the vector contains not more than max + 1 elements)
 * \return Vector with pointers to the start of each substring (empty substrings are omitted)
 * \note You have to free each element of the result vector!
 */
Vector<const char *> split(const char * source, const char * delimiter, size_t max = SIZE_MAX);
}  // namespace String
