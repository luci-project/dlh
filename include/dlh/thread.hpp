#pragma once

#include <dlh/utility.hpp>
#include <dlh/systypes.hpp>

struct Thread {
	 /* Pointer to the Thread Control Block */
	Thread * tcb = nullptr;
	/* Dynamic Thread Vector for Thread Local Storage */
	union DynamicThreadVector {
		size_t counter;
		struct dtv_pointer {
			void *val;
			void *to_free;
		} pointer;

		constexpr DynamicThreadVector(void * val = nullptr) : pointer{val, 0} {}

		inline bool allocated() const {
			return pointer.val != nullptr;
		}
	} * dtv = nullptr;
	/* Pointer to the thread descriptor.  */
	Thread * selfptr  = nullptr;
	int multiple_threads = 0;
	int gscope_flag = 0;
	uintptr_t sysinfo = 0;
	uintptr_t stack_guard = 0;
	uintptr_t pointer_guard = 0;
	unsigned long int vgetcpu_cache[2] = { 0, 0};
	unsigned int feature_1 = 0;
	int __glibc_unused1 = 0;
	/* Reservation of some values for the TM ABI.  */
	void *__private_tm[4] = { nullptr };
	/* GCC split stack support.  */
	void *__private_ss  = { nullptr };
	/* The lowest address of shadow stack,  */
	unsigned long long int ssp_base = 0;
	/* Must be kept even if it is no longer used by glibc since programs,
	  like AddressSanitizer, depend on the size of tcbhead_t.  */
	void *__glibc_unused2[64] __attribute__ ((aligned (32))) = { nullptr };

	union {
		struct {
			// Custom fields (in reserved space)
			int tls_errno;
			uintptr_t map_base;
			size_t map_size;
		};
		void * __padding[8] = { nullptr };
	};

	/* This descriptor's link on the `stack_used' or `__stack_user' list.  */
	struct ListHead {
		ListHead *next = nullptr;
		ListHead *prev = nullptr;
	} list;
	/* Thread ID - which is also a 'is this thread descriptor (and therefore stack) used' flag.  */
	pid_t tid = -1;
	/* Ununsed.  */
	pid_t pid_ununsed = 0;

	/* List of robust mutexes the thread is holding.  */
	void *robust_prev = nullptr;
	struct {
		void *list = nullptr;
		long int futex_offset = 0;
		void *list_op_pending = nullptr;
	} robust_head;
	/* List of cleanup buffers.  */
	void *cleanup = nullptr;
	/* Unwind information.  */
	void *cleanup_jmp_buf = nullptr;
	/* Flags determining processing of cancellation.  */
	int cancelhandling = 0;
	/* Flags.  Including those copied from the thread attribute.  */
	int flags = 0;
	/* We allocate one block of references here.  This should be enough
	 to avoid allocating any memory dynamically for most applications.  */
	struct {
		/* Sequence number.  We use uintptr_t to not require padding on
		   32- and 64-bit machines.  On 64-bit machines it helps to avoid
		   wrapping, too.  */
		uintptr_t seq = 0;
		/* Data pointer.  */
		void *data = nullptr;
	} specific_1stblock[32];
	/* Two-level array for the thread-specific data.  */
	void * specific[32] = { nullptr };
	/* Flag which is set when specific data is set.  */
	bool specific_used = false;
	/* True if events must be reported.  */
	bool report_events = false;
	/* True if the user provided the stack.  */
	bool user_stack = false;
	/* True if thread must stop at startup time.  */
	bool stopped_start = false;
	/* The parent's cancel handling at the time of the pthread_create
	 call.  This might be needed to undo the effects of a cancellation.  */
	int parent_cancelhandling = 0;
	/* Lock to synchronize access to the descriptor.  */
	int lock = 0;
	/* Lock for synchronizing setxid calls.  */
	unsigned int setxid_futex = 0;
	unsigned long long cpuclock_offset_ununsed = 0;
	/* If the thread waits to join another one the ID of the latter is
	   stored here.
	In case a thread is detached this field contains a pointer of the
	TCB if the thread itself.  This is something which cannot happen
	in normal operation.  */
	Thread * joindid = nullptr;
	/* The result of the thread function.  */
	void *result = nullptr;
	/* Scheduling parameters for the new thread.  */
	struct {
		int sched_priority  = 0;
	} schedparam;
	int schedpolicy = 0;
	/* Start position of the code to be executed and the argument passed to the function.  */
	void *(*start_routine) (void *) = nullptr;
	void *arg = nullptr;

	/* Debug state.  */
	struct {
		uint64_t eventmask = 0;          /* Mask of enabled events.  */
		int eventnum = 0;                /* Number of last event.  */
		void *eventdata = nullptr;       /* Data associated with event.  */
	} eventbuf;
	/* Next descriptor with a pending event.  */
	Thread *nextevent = nullptr;
	/* Machine-specific unwind info.  */
	struct {
		unsigned long exception_class = 0;
		void *exception_cleanup = nullptr;
		unsigned long  private_1 = 0;
		unsigned long  private_2 = 0;
	} exc;
	/* If nonzero, pointer to the area allocated for the stack and guard. */
	void *stackblock = nullptr;
	/* Size of the stackblock area including the guard.  */
	size_t stackblock_size = 0;
	/* Size of the included guard area.  */
	size_t guardsize = 0;
	/* This is what the user specified and what we will report.  */
	size_t reported_guardsize = 0;
	/* Thread Priority Protection data.  */
	void *tpp = nullptr;
	/* Resolver state.  */
	void * res[71] = { nullptr };
	/* Indicates whether is a C11 thread created by thrd_creat.  */
	bool c11 = false;
	/* This member must be last.  */
	char end_padding[0] = { };

	constexpr Thread(DynamicThreadVector * dtv = nullptr, uintptr_t base = 0, size_t size = 0, bool detach = false) : tcb(this), dtv(dtv), selfptr(this), map_base(base), map_size(size), joindid(detach ? this : 0) {}

	static Thread * create(void* (*func)(void*), void * arg = nullptr, bool detach = false, size_t stack_size = 1048576, size_t tls_size = 0xd40, DynamicThreadVector * dtv = nullptr, bool hidden = false);

	pid_t id() const {
		return tid;
	}

	bool detached() const {
		return joindid == this;
	}

	void detach();

	void exit(void * result = 0);

	int kill(signal_t sig);

	bool join(void** result = nullptr);

	void setup_guards(void* random);

	inline bool active() const {
		return tid > 0;
	}

	static inline Thread * self() {
		struct Thread *s;
		asm volatile("mov %%fs:%c1,%0"
		             : "=r" (s)
		             : "i" (__builtin_offsetof(Thread, selfptr))
		);
		return s;
	}

	static inline int & local_errno() {
		return self()->tls_errno;
	}
} __attribute__((aligned(64)));

static_assert(sizeof(Thread) == 0x900, "Wrong [P]Thread size");
