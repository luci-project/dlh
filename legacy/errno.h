#pragma once

#include <dlh/types.hpp>
#include <dlh/error.hpp>

#ifndef errno
#define errno           (*__errno_location())
#endif

extern "C" int *__errno_location() __attribute__((weak));

extern "C" char * strerror_r(int errnum, char *buf, size_t buflen) __attribute__((weak));
extern "C" char * strerror(int errnum) __attribute__((weak));

extern "C" void perror(const char * msg) __attribute__((weak));
