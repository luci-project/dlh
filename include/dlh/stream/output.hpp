#pragma once

#include <dlh/syscall.hpp>
#include <dlh/stream/string.hpp>

template<size_t BUFFERSZ>
class OutputStream : public StringStream<BUFFERSZ> {
 protected:
	int fd;

 public:
	/*! \brief Default constructor  */
	explicit OutputStream(int fd) : fd(fd) {}

	/*! \brief Destructor */
	virtual ~OutputStream() {}

	/*! \brief Flush the buffer.
	 */
	virtual void flush() override {
		size_t l = 0;
		while (l < this->_pos) {
			auto r = Syscall::write(fd, this->_bufptr + l, this->_pos - l);
			if (!r.valid())
				break;
			else
				l += r.value();
		}
		this->_pos = 0;
	}

};

extern OutputStream<1024> cout, cerr;
