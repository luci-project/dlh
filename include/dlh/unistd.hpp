#pragma once

// Based on musl libc (and rarely glibc)
#include <dlh/errno.hpp>
#include <dlh/types.hpp>

// Architecture-specific state (arch_prctl)
typedef enum : int {
	ARCH_SET_GS = 0x1001,
	ARCH_SET_FS = 0x1002,
	ARCH_GET_FS = 0x1003,
	ARCH_GET_GS = 0x1004
} arch_code_t;

// access
enum : int {
	F_OK = 0,
	X_OK = 1,
	W_OK = 2,
	R_OK = 4
};

// futex
struct timespec {
	long int tv_sec, tv_nsec;
};

typedef enum : int {
	FUTEX_WAIT            =  0,
	FUTEX_WAKE            =  1,
	FUTEX_FD              =  2,
	FUTEX_REQUEUE         =  3,
	FUTEX_CMP_REQUEUE     =  4,
	FUTEX_WAKE_OP         =  5,
	FUTEX_LOCK_PI         =  6,
	FUTEX_UNLOCK_PI       =  7,
	FUTEX_TRYLOCK_PI      =  8,
	FUTEX_WAIT_BITSET     =  9,
	FUTEX_WAKE_BITSET     = 10,
	FUTEX_WAIT_REQUEUE_PI = 11,
	FUTEX_CMP_REQUEUE_PI  = 12
} futex_op_t;

// Process ID function (get*id)
typedef int pid_t;

// Ressource limit (getrlimit)
struct rlimit {
	unsigned long rlim_cur;
	unsigned long rlim_max;
};


// File function (f*)
typedef long int off_t;
#define O_RDONLY           00
#define O_WRONLY           01
#define O_RDWR             02
#define O_CREAT          0100
#define O_EXCL           0200
#define O_TRUNC         01000
#define O_APPEND        02000
#define O_NONBLOCK      04000

#define O_SYNC       04010000
#define O_ASYNC        020000
#define O_LARGEFILE   0100000
#define O_DIRECTORY   0200000
#define O_NOFOLLOW    0400000
#define O_CLOEXEC    02000000
#define O_DIRECT       040000
#define O_NOATIME    01000000
#define O_PATH      010000000
#define O_DSYNC        010000
#define O_TMPFILE  (020000000 | __O_DIRECTORY)

// (f*at)
#define AT_FDCWD                -100
#define AT_SYMLINK_NOFOLLOW    0x100
#define AT_REMOVEDIR           0x200
#define AT_SYMLINK_FOLLOW      0x400
#define AT_NO_AUTOMOUNT        0x800
#define AT_EMPTY_PATH         0x1000
#define AT_STATX_SYNC_TYPE    0x6000
#define AT_STATX_SYNC_AS_STAT 0x0000
#define AT_STATX_FORCE_SYNC   0x2000
#define AT_STATX_DONT_SYNC    0x4000
#define AT_RECURSIVE          0x8000
#define AT_EACCESS             0x200

// (fcntl)
typedef enum : int {
	F_DUPFD            =    0,
	F_GETFD            =    1,
	F_SETFD            =    2,
	F_GETFL            =    3,
	F_SETFL            =    4,
	F_SETOWN           =    8,
	F_GETOWN           =    9,
	F_SETSIG           =   10,
	F_GETSIG           =   11,
	F_SETOWN_EX        =   15,
	F_GETOWN_EX        =   16,
	F_SETLEASE         = 1024,
	F_GETLEASE         = 1025,
	F_NOTIFY           = 1026,
	F_SETPIPE_SZ       = 1031,
	F_GETPIPE_SZ       = 1032,
	F_ADD_SEALS        = 1033,
	F_GET_SEALS        = 1034,
	F_GET_RW_HINT      = 1035,
	F_SET_RW_HINT      = 1036,
	F_GET_FILE_RW_HINT = 1037,
	F_SET_FILE_RW_HINT = 1038,
	F_DUPFD_CLOEXEC    = 1030,
} fcntl_cmd_t;

#define FD_CLOEXEC 1
#define F_RDLCK  0
#define F_WRLCK  1
#define F_UNLCK  2

#define F_SEAL_SEAL         0x0001
#define F_SEAL_SHRINK       0x0002
#define F_SEAL_GROW         0x0004
#define F_SEAL_WRITE        0x0008
#define F_SEAL_FUTURE_WRITE 0x0010

// File stat (*stat)
struct stat {
	unsigned long   st_dev;
	unsigned long   st_ino;
	unsigned        st_mode;
	unsigned long   st_nlink;
	unsigned        st_uid;
	unsigned        st_gid;
	unsigned long   st_rdev;
	off_t           st_size;
	long            st_blksize;
	long            st_blocks;
	struct timespec st_atim;
	struct timespec st_mtim;
	struct timespec st_ctim;
};

// Memory Mapping (m*)
#define PROT_NONE       0x0
#define PROT_READ       0x1
#define PROT_WRITE      0x2
#define PROT_EXEC       0x4

#define PROT_GROWSDOWN  0x01000000
#define PROT_GROWSUP    0x02000000

#define MAP_FAILED    ((void*)-1)

#define MAP_SHARED    0x01
#define MAP_PRIVATE   0x02
#define MAP_FIXED     0x10
#define MAP_ANONYMOUS 0x20

#define MS_ASYNC      1
#define MS_INVALIDATE 2
#define MS_SYNC       4

// Signals
typedef enum : int {
	SIGHUP    =  1,
	SIGINT    =  2,
	SIGQUIT   =  3,
	SIGILL    =  4,
	SIGTRAP   =  5,
	SIGABRT   =  6,
	SIGIOT    =  6,
	SIGBUS    =  7,
	SIGFPE    =  8,
	SIGKILL   =  9,
	SIGUSR1   = 10,
	SIGSEGV   = 11,
	SIGUSR2   = 12,
	SIGPIPE   = 13,
	SIGALRM   = 14,
	SIGTERM   = 15,
	SIGSTKFLT = 16,
	SIGCHLD   = 17,
	SIGCONT   = 18,
	SIGSTOP   = 19,
	SIGTSTP   = 20,
	SIGTTIN   = 21,
	SIGTTOU   = 22,
	SIGURG    = 23,
	SIGXCPU   = 24,
	SIGXFSZ   = 25,
	SIGVTALRM = 26,
	SIGPROF   = 27,
	SIGWINCH  = 28,
	SIGIO     = 29,
	SIGPOLL   = 29,
	SIGPWR    = 30,
	SIGSYS    = 31,
	SIGUNUSED = 31,
} signal_t;

// Signal handling
#define SIG_BLOCK          0	/* for blocking signals */
#define SIG_UNBLOCK        1	/* for unblocking signals */
#define SIG_SETMASK        2	/* for setting the signal mask */




// Syscall Wrapper
extern "C" pid_t gettid();
extern "C" pid_t getpid();
extern "C" pid_t getppid();

extern "C" int getrlimit(int resource, struct rlimit *rlim);
extern "C" int arch_prctl(arch_code_t code, unsigned long addr);

extern "C" int raise(signal_t sig);
extern "C" [[noreturn]] void abort();
extern "C" [[noreturn]] void exit(int code);

extern "C" void *mmap(void *start, size_t len, int prot, int flags, int fd, long off);
extern "C" int mprotect(void *addr, size_t len, int prot);
extern "C" int munmap(void *start, size_t len);
extern "C" int msync(void *start, size_t len, int flags);

extern "C" int access(const char *filename, int amode);
extern "C" int open(const char *filename, int flags);
extern "C" ssize_t read(int fd, void *buf, size_t count);
extern "C" ssize_t write(int fd, const void *buf, size_t size);
extern "C" int close(int fd);
extern "C" int fcntl(int fd, fcntl_cmd_t cmd, unsigned long arg = 0);
extern "C" int fallocate(int fd, int mode, off_t base, off_t len);
extern "C" int ftruncate(int fd, off_t length);

extern "C" int fstat(int fd, struct stat *st);
extern "C" int stat(const char * __restrict__ path, struct stat * __restrict__ buf);
extern "C" int lstat(const char * __restrict__ path, struct stat * __restrict__ buf);

extern "C" int memfd_create(const char *name, unsigned flags);

extern "C" int inotify_init();
extern "C" int inotify_init1(int flags);
extern "C" int inotify_add_watch(int fd, const char *pathname, uint32_t mask);
extern "C" int inotify_rm_watch(int fd, int wd);

extern "C" int futex(int * __restrict__ uaddr, futex_op_t futex_op, int val, const void * __restrict__ timeout, int * __restrict__ uaddr2, int val3);

extern "C" int brk(void * addr) __attribute__((weak));
extern "C" void * sbrk(intptr_t inc) __attribute__((weak));
