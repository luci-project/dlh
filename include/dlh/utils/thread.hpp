#pragma once

#include <dlh/utility.hpp>
#include <dlh/unistd.hpp>

struct Thread {
	// Thread Control Block
	Thread * tcb;
	union DynamicThreadVector {
		size_t counter;
		struct dtv_pointer {
			void *val;
			void *to_free;
		} pointer;

		constexpr DynamicThreadVector(void * val = reinterpret_cast<void*>(-1)) : pointer{val, 0} {}

		inline bool allocated() const {
			return pointer.val != reinterpret_cast<void*>(-1);
		}
	} * dtv;
	Thread * selfptr;
	int multiple_threads = 1;
	int gscope_flag = 0;
	uintptr_t sysinfo = 0;
	uintptr_t stack_guard = 0x2badc0de;
	uintptr_t pointer_guard = 0x1badc0de;
	void * __padding[77] = { nullptr };

	// Custom fields (in reserved space)
	bool detached = false;
	int tls_errno = 0;
	int exit_code = 0;
	pid_t tid = -1;
	void *map_base;
	size_t map_size;

	constexpr Thread(DynamicThreadVector * dtv = nullptr, void * base = nullptr, size_t size = 0, bool detach = false) : tcb(this), dtv(dtv), selfptr(this), detached(detach), map_base(base), map_size(size) {}

	static Thread * create(int (*func)(void*), void * arg = nullptr, bool detach = false, size_t stack_size = 1048576, size_t tls_size = 0xd40, DynamicThreadVector * dtv = nullptr, bool hidden = false);

	pid_t id() const {
		return tid;
	}

	void detach();

	void exit(int ret = 0);

	int kill(signal_t sig);

	bool join(int & exit_code);

	inline bool join() {
		int e;
		return join(e);
	}


	inline bool active() const {
		return tid > 0;
	}

	static inline Thread * self() {
		struct Thread *s;
		asm volatile("mov %%fs:%c1,%0"
		             : "=r" (s)
		             : "i" (__builtin_offsetof(Thread, selfptr))
		);
		return s;
	}

	static inline int & local_errno() {
		return self()->tls_errno;
	}
};

static_assert(sizeof(Thread) == 0x2c0, "Wrong Thread[ControlBlock] size");
