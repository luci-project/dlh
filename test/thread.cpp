#include <dlh/stream/output.hpp>
#include <dlh/utils/thread.hpp>
#include <dlh/unistd.hpp>
#include <dlh/errno.hpp>
#include <dlh/string.hpp>

void term(int signum) {
	write(2, "thread quit\n", 13);
	exit(0);
}

int foo(void * a) {
cout << "start " << __errno_location() << endl;
	cout << "int foo(" << a << ")" << endl;
	const char * name = "observer";
	errno = 0;

	struct sigaction action;
	memset(&action, 0, sizeof(struct sigaction));
	action.sa_handler = term;
	auto sr = sigaction(SIGCHLD, &action, NULL);
	if (sr == -1)
		perror("sigaction");

cout << "start" << endl;

	auto r = prctl(PR_SET_PDEATHSIG, SIGCHLD);
	if (r == -1)
		perror("Set signal");
	cout << "Set signal " << r << endl;


	sleep(8);
	cout << "foo done" << endl;
	return 42;
}

int main(int argc, const char *argv[]) {
	cerr << "start " << __errno_location() << endl;
	Thread * f = Thread::create(foo, (void*)0x23, true);
	if (f == nullptr) {
		cerr << "failed" << endl;
		return 1;
	}
	sleep(4);
	cerr << "main exit" << endl;
	return 23;
}
