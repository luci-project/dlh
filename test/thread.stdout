foo start
int foo(23)
set signal 0
